### PR TITLE
Wait to raise Connected event for Read connection until container is caught up

### LINKED ISSUE #7884: Connectivity eventing improvements
Per https://office.visualstudio.com/OC/_workitems/edit/5489692?src=WorkItemMention&src-action=artifact_link, we need to
- ~expose "connect" event from Container, to allow hosts to know when connection has been established (and stop showing disconnected banner).~
  - [update] -- @markfields will work with @scottn12 to clarify the ideal list of events, including updated names and very clear semantics/gaurantees. The plan is written up in https://github.com/microsoft/FluidFramework/issues/9677.
- Change when "connected" event is raised for "read" connections - make it fire only when container is up-to-date (based on what we know, i.e. checkpointSequenceNumber, maybe storage call coming back)
   - This will ensure that transition from "read" to "write" connection will not suddenly make container not up-to-date.

### PR DESCRIPTION
**Pull request by markfields**
Part of #7884.  See comments in connectionStateHandler.ts for description of the desired behavior here.

Note that the new behavior is behind a feature gate "Fluid.Container.CatchUpBeforeDeclaringConnected".  I didn't make any changes to existing function `waitContainerToCatchUp`, even though some functionality is duplicated.  Once we finish rolling this out, we can remove the duplicate logic from that function since it will simply wait for the "connected" event (we may even deprecate it at that point)

To-Do:
- [ ] Need to update the README and specifically info about CatchingUp state

Some follow-up work items to open afterwards:
- [ ] Roll out new feature gate to enable new behavior for Loop partners
- [ ] Remove feature gate and deprecate `waitContainerToCatchUp`

### CODE REVIEW

File: packages/loader/container-loader/src/container.ts
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:200]: Looks like we never clear this callback on successful path - might be worth fixing it.

File: packages/loader/container-loader/src/container.ts
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:218]: can you please remove eslint comment below?

File: packages/loader/container-loader/src/container.ts
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:202]: I think it points out to a fact that if we do not have checkpointSequenceNumber, then we can at least leverage join op sequence number for same purposes (for "write" connections).

What this code does not do a good job is dealing with disconnects while waiting for last op. I think it's worth opening a bug to incorporate that.

**Comment by vladsud** [comment]
I think it looks good. Some things missing:
1. Tests
2. breaking.md note
3. Possibly: some feature gate to disable that change, in case we learn it screws up something we did not consider.

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (agarwal-navin) [inline packages/loader/container-loader/src/connectionStateHandler.ts:197]: I would call this `clientAlreadyInQuorum` and add a comment saying that if the client is in quorum, it would have processed it's own join op.

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (agarwal-navin) [inline packages/loader/container-loader/src/connectionStateHandler.ts:206]: Based on the if condition above, this may not be true right. We can be in this else block if the connection is write and we haven't processed our join op.

Comment (agarwal-navin) [reply packages/loader/container-loader/src/connectionStateHandler.ts:206]: Ah okay, my bad. In the scenario I mentioned, we will go in the if condition. We will be here if it's a write connection and we have processed our join op.

Comment (agarwal-navin) [reply packages/loader/container-loader/src/connectionStateHandler.ts:206]: But this still may not be immediate. There could have been more ops after the join op and in that case, we will wait for those ops to be processed before we move to connected state.

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (agarwal-navin) [inline packages/loader/container-loader/src/connectionStateHandler.ts:207]: I don't think "identical semantics to write connection" is entirely correct. It should explicitly mention what happens in read vs write. In both read and write, we wait to catch up to the sequence number we know about at this point in time. In write, we also wait for our own join op.
You should add this block to the big comment block before the if condition. And then maybe have comments in the if / else block specific to what happens in that scenario. Right now, its a little confusing.

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (agarwal-navin) [reply packages/loader/container-loader/src/connectionStateHandler.ts:206]: Also, please use `processed` our join op vs `seen` which mean different things.

File: packages/loader/container-loader/src/container.ts
Comment (agarwal-navin) [inline packages/loader/container-loader/src/container.ts:161]: `waitForOps` is too generic. Maybe this should be called `waitToProcessKnownOps`?

Comment (agarwal-navin) [reply packages/loader/container-loader/src/container.ts:161]: Or `catchUpToKnownOps`

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (agarwal-navin) [inline packages/loader/container-loader/src/connectionStateHandler.ts:21]: nit: `onCaughUpToKnownOps`.

File: packages/loader/container-loader/src/container.ts
Comment (markfields) [reply packages/loader/container-loader/src/container.ts:200]: Suggested @Abe27342 take care of this in PR https://github.com/microsoft/FluidFramework/pull/10071 since he's changing the callback.

File: packages/loader/container-loader/src/container.ts
Comment (markfields) [reply packages/loader/container-loader/src/container.ts:218]: Not going to touch it - `resume` has been deprecated and replaced with `connect`, @scottn12 will be replacing it soon (https://github.com/microsoft/FluidFramework/issues/9397)

File: packages/loader/container-loader/src/container.ts
Comment (markfields) [reply packages/loader/container-loader/src/container.ts:200]: And he did - yay.

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (markfields) [reply packages/loader/container-loader/src/connectionStateHandler.ts:206]: > There could have been more ops after the join op and in that case, we will wait for those ops to be processed before we move to connected state.

Yeah re-reading this change I came to the same conclusion, glad you saw it too, I was trying to figure out if I was missing something now or before when I wrote the code :P

**Comment by markfields** [comment]
Revisiting this finally.  Writing down some questions I'm trying to clarify in my head:

* How does `startJoinOpTimer` work?  When it fires, it doesn't do anything.  I guess `receivedAddMemberEvent` keys off whether it's set or not, and if so may call `applyForConnectedState` which in turn may call `this.setConnectionState(ConnectionState.Connected);`.
* For write connections, does `waitContainerToCatchUp` actually cover the scenario of leveraging join op to indicate "caught up"?  The comment in there (the one I said I don't understand) seems to indicate so, but I don't see it.

**Comment by markfields** [comment]
@jatgarg can you take a look too please? Thanks!

**Comment by markfields** [comment]
Answering my own question...

> For write connections, does waitContainerToCatchUp actually cover the scenario of leveraging join op to indicate "caught up"? The comment in there (the one I said I don't understand) seems to indicate so, but I don't see it.

`waitContainerToCatchUp` gates on `connected` event which for write connections means we've processed own join op.

This PR kind of inverts this - now _always_ "connected" event means we're caught up.  So `waitContainerToCatchUp` is actually "wait for the container to be connected".

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (markfields) [reply packages/loader/container-loader/src/connectionStateHandler.ts]: I made a bunch of changes to the comments, not sure I exactly addressed this but take a look and see if it needs further clarification?

**Comment by markfields** [comment]
@vladsud, @agarwal-navin, @jatgarg -- Here's my update on this PR, and some input I need from you:

* Do we all agree with the specification that "connected" means "caught up to known ops" even for write clients? i.e. that's an additional delay beyond the "join" op now.  Also, the way I've implemented so far the definition of "known ops" varies based on when a particular callback is invoked, so it's a little wishy-washy.  Thoughts?
* I started looking into a feature gate for this.  Ideally the gate would be used in two places: (1) ConnectionStateHandler to determine when to transition to "connected" state and (2) waitContainerToCatchUp can just wait for "connected" event when the gate is enabled (don't need to wait for additional ops).  I don't see a way to cover (2) because it's a function taking in an `IContainer`, rather than being a private function on `Container` class.  I think this will be fine for `waitContainerToCatchUp` to essentially catch up twice - if `hasCheckpointSequenceNumber` it'll be a no-op the second time anyway.
  * If we agree then I'll revert behavior of `waitContainerToCatchUp` back to how it was.
* I need to write tests
* I need to move this to `next` and include BREAKING.md note.

Anything else I should consider here?  I anticipate moving this from Draft on Monday or Tuesday.  Thanks guys!

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (vladsud) [inline packages/loader/container-loader/src/connectionStateHandler.ts:132]: I do not think this is correct flow. In a sense that it would work only if we do not lose connection while processing ops.
If we do, we have to abandon this process and start over.

File: packages/loader/container-loader/src/container.ts
Comment (jatgarg) [inline packages/loader/container-loader/src/container.ts:616]: nit: onCaughUpToKnownOps -> onCaughtUpToKnownOps

Comment (jatgarg) [inline packages/loader/container-loader/src/container.ts:616]: The signature you added above for "callback" is different.

File: packages/loader/container-loader/src/container.ts
Comment (jatgarg) [inline packages/loader/container-loader/src/container.ts:203]: "usedCheckpointSequenceNumber " does not seem right, I think it should just be "hasCheckpointSequenceNumber"

File: packages/loader/container-loader/src/container.ts
Comment (markfields) [reply packages/loader/container-loader/src/container.ts:616]: Yeah this code doesn't care about the boolean parameter in the callback so just omit it.  The function types are compatible (the argument will just be dropped by JS).

**Comment by markfields** [comment]
@agarwal-navin @jatgarg ready for Review - Except I will rebase onto next, and I will write a bunch of tests.  Thanks!

File: packages/loader/container-loader/src/catchUpMonitor.ts
Comment (tyler-cai-microsoft) [inline packages/loader/container-loader/src/catchUpMonitor.ts:18]: What does the catchUpMonitor monitor catching up to?

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (tyler-cai-microsoft) [inline packages/loader/container-loader/src/connectionStateHandler.ts:36]: `JoinOpTimeoutMs`

File: packages/loader/container-loader/src/catchUpMonitor.ts
Comment (markfields) [reply packages/loader/container-loader/src/catchUpMonitor.ts:18]: I'll add a comment, thanks

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (tyler-cai-microsoft) [inline packages/loader/container-loader/src/connectionStateHandler.ts:275]: I'm assuming these are do nothing calls?

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (tyler-cai-microsoft) [inline packages/loader/container-loader/src/connectionStateHandler.ts:258]: Is it worth turning this magic string into an exported variable?

File: packages/loader/container-loader/src/catchUpMonitor.ts
Comment (tyler-cai-microsoft) [inline packages/loader/container-loader/src/catchUpMonitor.ts:68]: where does `caughtUpListener` come from?

**Comment by markfields** [comment]
Remaining tasks:
- [ ] Write UTs for CatchUpMonitor and ImmediateCatchUpMonitor
- [ ] Fix the skipped test in Container.spec.ts
- [ ] Address //* comments

File: packages/loader/container-loader/src/container.ts
Comment (markfields) [inline packages/loader/container-loader/src/container.ts:182]: What do you all think - is this correct or not?  I may not add additional delay since events fire synchronously.

**Comment by markfields** [comment]
@agarwal-navin @vladsud @tyler-cai-microsoft @jatgarg -- Thanks for the early review!  Ready for review now.  Just need to add UTs for the new "CatchUpMonitor" classes but otherwise ready to go.

File: packages/loader/container-loader/src/container.ts
Comment (jatgarg) [inline packages/loader/container-loader/src/container.ts:1516]: Have you given a thought that it should be opt out and not opt in?

Comment (markfields) [reply packages/loader/container-loader/src/container.ts:1516]: Only a little.  I think shipping dark is the standard operating procedure, and there's no rush to diverge from this.  Also I personally want to practice what it's like to use feature gates.

File: packages/loader/container-loader/src/catchUpMonitor.ts
Comment (markfields) [reply packages/loader/container-loader/src/catchUpMonitor.ts:62]: It's the callback provided by whoever is calling `monitor.on("caughtUp")`  This is a trick so we can hide the `caughtUp` state - the other approach is anyone who ever registers for "caughtUp" needs to first check to see if we already are.

File: packages/loader/container-loader/src/test/connectionStateHandler.spec.ts
Comment (markfields) [inline packages/loader/container-loader/src/test/connectionStateHandler.spec.ts:114]: Note to self - delete this block

File: packages/loader/container-loader/src/container.ts
Comment (jatgarg) [inline packages/loader/container-loader/src/container.ts:1517]: nit: No need to check "=== true"

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (jatgarg) [inline packages/loader/container-loader/src/connectionStateHandler.ts:238]: We are moving towards "relay service" terminology than "ordering service". So, we should revert this change.

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (jatgarg) [inline packages/loader/container-loader/src/connectionStateHandler.ts:244]: nit: "Connected event" -> "Connected state"

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (jatgarg) [inline packages/loader/container-loader/src/connectionStateHandler.ts:230]: Shouldn't this be an invariant and should we throw on it?

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (jatgarg) [inline packages/loader/container-loader/src/connectionStateHandler.ts:284]: Same as above.

File: packages/loader/container-loader/src/catchUpMonitor.ts
Comment (jatgarg) [inline packages/loader/container-loader/src/catchUpMonitor.ts:84]: Cann't we just add the type of listener here? instead of typecasting later?

File: packages/loader/container-loader/src/catchUpMonitor.ts
Comment (jatgarg) [inline packages/loader/container-loader/src/catchUpMonitor.ts:55]: I think we should change the order of statements here on line 52 and 55. So like initialize first before registering the handler.

File: packages/loader/container-loader/src/catchUpMonitor.ts
Comment (jatgarg) [inline packages/loader/container-loader/src/catchUpMonitor.ts:62]: You are passing "transitionToConnectedState" as "caughtUpListener", but transitionToConnectedState does not take any argument. So, what is the logic here of passing the "hasCheckpointSequenceNumber".

Comment (markfields) [reply packages/loader/container-loader/src/catchUpMonitor.ts:62]: I went back and forth on this.  In the future when this feature gate is at 100%, `waitContainerToConnect` will become obsolete, replaced by simply waiting for the `connected` event.  If someone were depending on the boolean return value of that function, maybe we would want to update `connected` to include that in its event args.

I feel it's kind of a strange thing to expose, I guess the point is some kind of "confidence" indicator of how caught up you'll be when the thing resolves/fires.  @vladsud thoughts on maintaining this info long-term (and whether it belongs in the short term)

File: packages/loader/container-loader/src/catchUpMonitor.ts
Comment (markfields) [reply packages/loader/container-loader/src/catchUpMonitor.ts:84]: I like it as-is because if there were other events supported here the listener would be one of many types (probably a union type) and then you'd need to narrow it down depending on the event.

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (markfields) [reply packages/loader/container-loader/src/connectionStateHandler.ts:297]: No need, It's protected by the type system 😎

File: packages/loader/container-loader/src/container.ts
Comment (markfields) [reply packages/loader/container-loader/src/container.ts:1517]: I prefer to be explicit that it must be set to true (but you're right the code works either way)

File: packages/loader/container-loader/src/catchUpMonitor.ts
Comment (vladsud) [inline packages/loader/container-loader/src/catchUpMonitor.ts:58]: It feels like instead of using events, it's better to use a property that exposes a promise. I.e. something like
```ts
public caughtUp: Promise<number>;
```
Promise seems to be better because this event files once, and past "firing" reflected on new listeners - that's exactly what promise does.

File: packages/loader/container-loader/src/catchUpMonitor.ts
Comment (vladsud) [reply packages/loader/container-loader/src/catchUpMonitor.ts:62]: I'm fine not exposing this info anywhere. It used to be that FRS was an outlier here as it did not have checkpoint sequence number. But with recent work to expose sequence numbers on join & leave signals, we should ensure that both FRS is on par with ODSP, as this feature is just too good not to have it - it allows many things beyond ability to catch up, like building sync primitives (including leader / summarizer election) using signals only. Plus, it was added to solve race condition on connection that resulted in client missing ops - so it's really not that optional to have it.

File: packages/loader/container-loader/src/catchUpMonitor.ts
Comment (markfields) [reply packages/loader/container-loader/src/catchUpMonitor.ts:84]: This is just a notification after the listener is added. It'll get removed when `dispose` is called below which happens on disconnect.

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (markfields) [reply packages/loader/container-loader/src/connectionStateHandler.ts:187]: For the record - Vlad prototyped this in #10958 and is thinking about doing the refactor after this PR is merged.

File: packages/loader/container-loader/src/connectionStateHandler.ts
Comment (markfields) [reply packages/loader/container-loader/src/connectionStateHandler.ts:223]: Prototype of this direction is in #10958 - thanks for pursuing this, definitely an improvement!

**Review by vladsud — APPROVED** [review]
Singing off to unblock you - I did not look through it now, but was poking in my prototyping on top of it and overall, it looks fine.

File: packages/loader/container-loader/src/catchUpMonitor.ts
Comment (markfields) [reply packages/loader/container-loader/src/catchUpMonitor.ts:56]: That is a great point.  I don't remember why I switched to event-based to begin with, but I bet it would work better with Promise.  That said, I don't have time to try it out now, and also leaving it will reduce merge conflicts with your follow-on refactor.

File: packages/loader/container-loader/src/catchUpMonitor.ts
Comment (markfields) [reply packages/loader/container-loader/src/catchUpMonitor.ts]: getting rid of hasCheckpointSequenceNumber makes it a little more generic, I'll see about adding some comments or any other asserts etc

Comment (markfields) [reply packages/loader/container-loader/src/catchUpMonitor.ts]: I looked at adding `deltaManager.on("disconnect", () => this.dispose())`, which I do like.  I am sorry I am just out of time though.  I would want to add tests, etc.  It's an easy change to tack on later.

Also, I think it works across connections anyway.

**Comment by markfields** [comment]
Opened https://dev.azure.com/fluidframework/internal/_workitems/edit/1384 to track rolling this out

**Pull request merged**
